### PR TITLE
Noise scaling update

### PR DESCRIPTION
**Pull request by TorkelE**
Updates the noise scaling feature with:
- Ensures that species/variables/parameters introduced by the noise scaling metadata is always properly added to the new system.
- Add additional tests.

### CODE REVIEW

File: src/reactionsystem.jl [hunk 1]
```diff
@@ -1221,7 +1222,34 @@ function set_default_metadata(rs::ReactionSystem;  default_reaction_metadata = [
...
     @set! rs.rxs = Reaction[rx for rx in updated_equations if rx isa Reaction]
-    
+
+    # Special routine to handle `Reaction` metadata that can contain new symbolic variables.
+    # Currently, `noise_scaling` is the only relevant metadata supported this way.
+    drm_dict = Dict(default_reaction_metadata)
+    if haskey(drm_dict, :noise_scaling)
+        # Finds parameters, species, and variables in the noise scaling term.
+        ns_expr = drm_dict[:noise_scaling]
+        ns_syms = [Symbolics.unwrap(sym) for sym in get_variables(ns_expr)]
```
Comment (isaacsas) [inline src/reactionsystem.jl:1232]: `unwrapping` should be handled in the `Reaction` constructor. We shouldn't be storing things that aren't unwrapped as we don't want to have to think about this internally later on (at least every internal Catalyst function I've written assumes its input is unwrapped as this is the MTK convention).

File: src/reactionsystem.jl [hunk 2]
```diff
@@ -1221,7 +1222,34 @@ function set_default_metadata(rs::ReactionSystem;  default_reaction_metadata = [
...
+            sps_new = unique([get_species(rs); ns_sps])
+            vs_old = get_unknowns(rs)[length(get_species(rs))+1 : end]
+            @set! rs.species = sps_new
+            @set! rs.unknowns = [sps_new; vs_old]
+        end
+        if any(!any(isequal(v1, v2) for v2 in get_unknowns(rs)) for v1 in ns_vs)
+            us_new = unique([get_unknowns(rs); ns_vs])
+            @set! rs.unknowns = us_new
+        end
+    end
```
Comment (isaacsas) [inline src/reactionsystem.jl:1251]: Does `union!` not work?

**Review by isaacsas — CHANGES_REQUESTED** [review]

File: src/reactionsystem.jl (hunk 1 above)
Comment (TorkelE) [reply src/reactionsystem.jl:1232]: Right now we just use the same convention for `Reaction` metadata as MTK/Symbolics does for symbolic variables (i.e. metadata is stored as they are provided by the user, if the user sets a metadata as `Num` it remain as a `Num`).

If you want we can add a specific routine in the `Reaction` constructor that checks if there is any `noise_scaling` metadata, and if it is `Num` unwraps it at that stage.

File: src/reactionsystem.jl (hunk 2 above)
Comment (TorkelE) [reply src/reactionsystem.jl:1251]: I never really knew about `union!` until I discovered that you use it in a few cases, so not that familiar with it. That said I should start using it, will implement it here.

Comment (TorkelE) [reply src/reactionsystem.jl:1251]: Yeah, I can probably use generators, I think at this stage it was late and I was getting increasingly annoyed at all the effort doing this extra step. Will update to use generators though.

Comment (TorkelE) [reply src/reactionsystem.jl:1251]: actually, I am a little bit uncertain exactly how you suggest we use `union!` and generators in these cases

Comment (isaacsas) [inline src/reactionsystem.jl:1251]:
```suggestion
        ns_syms = [Symbolics.unwrap(sym) for sym in get_variables(ns_expr)]
        ns_ps = Iterators.filter(ModelingToolkit.isparameter, ns_syms)
        ns_sps = Iterators.filter(Catalyst.isspecies, ns_syms)
        ns_vs = Iterators.filter(sym -> !Catalyst.isspecies(sym) && !ModelingToolkit.isparameter(sym), ns_syms)

        # Adds parameters, species, and variables to the `ReactionSystem`.
        @set! rs.ps = union(get_ps(rs), ns_ps])
         sps_new = union(get_species(rs), ns_sps)
         @set! rs.species = sps_new
         vs_old = @view get_unknowns(rs)[length(get_species(rs))+1 : end]            
         @set! rs.unknowns = union(sps_new, vs_old]
         @set! rs.unknowns = union(get_unknowns(rs), ns_vs)
        end       
    end
```
If this is acceptable and works / tests pass I'm fine with merging after switching to this.

**Pull request merged**